### ISSUE TITLE
Tokens not being processed with specifying user message as input parameter

### ISSUE THREAD

**Issue by edeandrea** [quarkiverse/quarkus-langchain4j#132]
If I supply a user message with tokens as an input parameter the tokens are not processed.

```java
@SystemMessage("You are a marvel comics writer, expert in all sorts of super heroes and super villains.")
String narrate(@UserMessage String userMessage, Fight fight);

default String narrate(Fight fight) {
    try {
        var userMessage = Files.readString(Path.of(getClass().getClassLoader().getResource("usermessage.txt").toURI()));
        return narrate(userMessage, fight);
    }
    catch (IOException | URISyntaxException e) {
        throw new RuntimeException(e);
    }
}
```

produces

```
In the heart of the bustling city, amidst towering skyscrapers and bustling streets, a fierce battle unfolded. The air crackled with anticipation as {fight.winnerName}, a formidable force for justice, faced off against {fight.loserName}, a cunning mastermind of chaos. The clash of powers was about to commence, and the fate of the city hung in the balance.

With a flash of lightning, {fight.winnerName} unleashed their incredible {fight.winnerPowers}, harnessing the elements themselves. The ground trembled beneath their feet as they summoned a mighty gust of wind, sending debris swirling through the air. Their level of mastery over their powers was unmatched, and it showed in every move they made.

But {fight.loserName} was no ordinary adversary. With a sly grin, they tapped into their own unique {fight.loserPowers}, manipulating shadows and illusions with a mischievous flair. Their level of trickery and deception was unparalleled, making it difficult for even the most astute heroes to discern reality from illusion.

As the battle raged on, the clash of powers illuminated the night sky, casting an ethereal glow over the city. {fight.winnerName} fought with unwavering determination, their powers reaching new heights as they tapped into their inner strength. With each strike, they pushed themselves further, refusing to let evil prevail.

In the end, it was {fight.winnerName} who emerged victorious. Their unwavering spirit and mastery over their {fight.winnerPowers} proved to be the deciding factor. With a final surge of energy, they delivered a decisive blow, incapacitating {fight.loserName} and restoring peace to the city once more. The citizens rejoiced, grateful for the unwavering heroism of {fight.winnerName} and the triumph of good over evil.

+++++
Name: {fight.winnerName}
Powers: {fight.winnerPowers}
Level: {fight.winnerLevel}
+++++

+++++
Name: {fight.loserName}
Powers: {fight.loserPowers}
Level: {fight.loserLevel}
+++++
```

where the contents of `usermessage.txt` is

```text
Narrate the fight between a super hero and a super villain.

During the narration, don't repeat "super hero" or "super villain".

Write 4 paragraphs maximum. Be creative.

The narration must be:
- G rated
- Workplace/family safe
- No sexism, racism, or other bias/bigotry

Here is the data you will use for the winner:

+++++
Name: {fight.winnerName}
Powers: {fight.winnerPowers}
Level: {fight.winnerLevel}
+++++

Here is the data you will use for the loser:

+++++
Name: {fight.loserName}
Powers: {fight.loserPowers}
Level: {fight.loserLevel}
+++++

Here is the data you will use for the fight:

+++++
{fight.winnerName} who is a {fight.winnerTeam} has won the fight against {fight.loserName} who is a {fight.loserTeam}.

The fight took place in {fight.location.name}, which can be described as {fight.location.description}.
+++++
```

whereas when I have this:

```java
  @SystemMessage("You are a marvel comics writer, expert in all sorts of super heroes and super villains.")
  @UserMessage("""
    Narrate the fight between a super hero and a super villain.

    During the narration, don't repeat "super hero" or "super villain".
    
    Write 4 paragraphs maximum. Be creative.
    
    The narration must be:
    - G rated
    - Workplace/family safe
    - No sexism, racism, or other bias/bigotry
    
    Here is the data you will use for the winner:
    
    +++++
    Name: {fight.winnerName}
    Powers: {fight.winnerPowers}
    Level: {fight.winnerLevel}
    +++++
    
    Here is the data you will use for the loser:
    
    +++++
    Name: {fight.loserName}
    Powers: {fight.loserPowers}
    Level: {fight.loserLevel}
    +++++
    
    Here is the data you will use for the fight:
    
    +++++
    {fight.winnerName} who is a {fight.winnerTeam} has won the fight against {fight.loserName} who is a {fight.loserTeam}.
    
    The fight took place in {fight.location.name}, which can be described as {fight.location.description}.
    +++++
    """)
  String narrate(Fight fight);
```

I get this:

```
In the gritty streets of Gotham City, a clash of epic proportions unfolded. Han Solo, a hero known for his sharpshooting skills and skepticism towards the force, faced off against Storm Trooper, a villain armed with nothing more than a small gun. The odds seemed stacked against the Storm Trooper, but he was determined to prove his worth.

As the battle commenced, Han Solo swiftly dodged the Storm Trooper's feeble shots, his agility and experience shining through. With a smirk on his face, Han Solo aimed his big gun with precision, firing shots that echoed through the city. The Storm Trooper stumbled, his small gun no match for the firepower of his opponent.

Undeterred, the Storm Trooper fought back with unwavering determination. He maneuvered through the chaos, attempting to outsmart Han Solo. But the hero's quick reflexes and strategic thinking proved to be too much for the Storm Trooper to handle. With each passing moment, Han Solo's level of expertise became more evident.

In a final, decisive move, Han Solo disarmed the Storm Trooper, leaving him defenseless and defeated. The hero's victory was celebrated by the citizens of Gotham City, who witnessed the triumph of justice over villainy. Han Solo, with his unwavering resolve and unmatched skills, had once again proven himself as a true hero in the face of adversity.
```

**Comment by geoand** [comment]
I'm having a hard time understanding what the issue is here. 

Can you elaborate please?

**Comment by geoand** [comment]
So the problem here is that in langchain4jm when you pass a string parameter, it's assumed that it is not a template. 

I see the following ways to support what you want:

* Add a way for the method `@UserMessage` and `@SystemMessage` annotations to result in reading data from a file. 
* Add an annotation like `@Template` that can be used on a method parameter that will force the evaluation of the parameter as a template.

**Comment by edeandrea** [comment]
Aha so I guess this is somewhat user/documentation error then.

When I read the docs I guess I just assumed that rather than hard-coding the user message I could pass it in and the behavior was the same.

[The docs](https://docs.quarkiverse.io/quarkus-langchain4j/dev/ai-services.html#_receiving_user_message_as_a_parameter) don't mention that the passed in user message can't be templated.

Ideally what I'd like to do is store the templated user message in a file and read it once to load the user message then reuse it. 

Maybe it's just me (yeah I know I'm a bit OCD :) ), but hard-coding a long string like that inside an annotation just seems wrong to me.

But regardless, if the passed in user message isn't expected to be a template, then the docs should say that. Otherwise it should support being templated.

**Comment by geoand** [comment]
> Ideally what I'd like to do is store the templated user message in a file and read it once to load the user message then reuse it.

Right, I think that's valid

**Comment by geoand** [comment]
I know you would not expect anything less :)

**Comment by edeandrea** [comment]
We'd first need to decide/agree upon what the proper functionality is/should be. Should a passed-in user message be allowed to be templated? I don't see why not to be honest.

I think being able to externalize the user/system messages (i.e. read from a file) would be an enhancement that I could create a separate issue for if you think those a valid use cases.

**Comment by geoand** [comment]
> We'd first need to decide/agree upon what the proper functionality is/should be

We will continue doing what upstream Langchain4j does. If this changes upstream, we can change here too.

**Comment by geoand** [comment]
> I think being able to externalize the user/system messages (i.e. read from a file) would be an enhancement that I could create a separate issue for if you think those a valid use cases.

Makes sense

**Comment by edeandrea** [comment]
> We will continue doing what upstream Langchain4j does. If this changes upstream, we can change here too.

Oh so upstream Langchain4j does not allow templatizing a passed in user message? I did not know that.

**Comment by geoand** [comment]
> Oh so upstream Langchain4j does not allow templatizing a passed in user message?

Yeah, a passed in parameter is itself considered a variable to the template that is otherwise known.

**Comment by edeandrea** [comment]
But the parameter has the `@UserMessage` annotation...

I can't find any documentation upstream that says whether or not templates if used as a method parameter are supported or not, nor do I have a sample project that is upstream only.

**Comment by geoand** [comment]
It should be easy to create an upstream project that does that based on what you see  in https://github.com/langchain4j/langchain4j-examples

**Comment by edeandrea** [comment]
It already has an example which does support a templated user message passed in as a method parameter:

https://github.com/langchain4j/langchain4j-examples/blob/aae3b1356057f10a6d2a12f618f7bc673dd7f35e/devoxx/src/main/java/_08_AIServiceExamples.java#L71

**Comment by edeandrea** [comment]
Actually I had to modify it a little bit:

```java
static class AI_Service_with_System_and_User_Messages_Example {

        interface TextUtils {

            @SystemMessage("You are a professional translator into {{language}}")
            @UserMessage("Translate the following text: {{text}}")
            String translate(@V("text") String text, @V("language") String language);

            @SystemMessage("Summarize every message from user in {{n}} bullet points. Provide only bullet points.")
            List<String> summarize(@UserMessage String text, @V("n") int n, @V("text") String t);
        }

        public static void main(String[] args) {

            TextUtils utils = AiServices.create(TextUtils.class, model);

            String translation = utils.translate("Hello, how are you?", "italian");
            System.out.println(translation); // Ciao, come stai?

            String text = "{{text}}, or artificial intelligence, is a branch of computer science that aims to create "
                    + "machines that mimic human intelligence. This can range from simple tasks such as recognizing "
                    + "patterns or speech to more complex tasks like making decisions or predictions.";

            List<String> bulletPoints = utils.summarize(text, 3, "AI");
            System.out.println(bulletPoints);
            // [
            // "- AI is a branch of computer science",
            // "- It aims to create machines that mimic human intelligence",
            // "- It can perform simple or complex tasks"
            // ]
        }
    }
```

**Comment by edeandrea** [comment]
But then this didn't work...

```java
static class AI_Service_with_System_and_User_Messages_Example {

        interface TextUtils {

            @SystemMessage("You are a professional translator into {{language}}")
            String translate(@UserMessage String userMessage, @V("text") String text, @V("language") String language);

            @SystemMessage("Summarize every message from user in {{n}} bullet points. Provide only bullet points.")
            List<String> summarize(@UserMessage String text, @V("n") int n, @V("text") String t);
        }

        public static void main(String[] args) {

            TextUtils utils = AiServices.create(TextUtils.class, model);

            String translation = utils.translate("Translate the following text: {{text}}", "Hello, how are you?", "italian");
            System.out.println(translation); // Ciao, come stai?

            String text = "{{text}}, or artificial intelligence, is a branch of computer science that aims to create "
                    + "machines that mimic human intelligence. This can range from simple tasks such as recognizing "
                    + "patterns or speech to more complex tasks like making decisions or predictions.";

            List<String> bulletPoints = utils.summarize(text, 3, "AI");
            System.out.println(bulletPoints);
            // [
            // "- AI is a branch of computer science",
            // "- It aims to create machines that mimic human intelligence",
            // "- It can perform simple or complex tasks"
            // ]
        }
    }
```

```
Translate the following text: {{text}}
[- Artificial intelligence is a branch of computer science, - AI aims to create machines that mimic human intelligence, - AI can perform tasks ranging from simple pattern recognition to complex decision-making]

Process finished with exit code 0

```

**Comment by geoand** [comment]
So AFAIR, langchain4j doesn't support having the method parameter `@UserMessage` being a template, correct?

**Comment by edeandrea** [comment]
> So AFAIU, langchain4j doesn't support having the method parameter `@UserMessage` being a template, correct?

Seems like it, unless things have changed since December.

**Comment by edeandrea** [comment]
Looks like this has been implemented upstream as of 0.30

https://github.com/langchain4j/langchain4j/pull/930

**Comment by geoand** [comment]
Interesting, thanks for the update
